### ISSUE TITLE
Interrupt safe queue API

### ISSUE THREAD

**Issue by quzhi010** [micro-os-plus/eclipse-test-projects#2]
Hi,

I've been trying to use this version on STM32F405 chip. I'm wondering whether it's safe to use the queue API within interrupt?

Based on the document, it seems like there's a there's a "try_send" method and I can use it to send message from low priority interrupt to a queue. It does work properly when the interrupt occurs at low frequency. However, when the interrupt fires at 3~5Hz, the "try_send" method would just stop working at some point.

I tried to use a volatile variable to count how many messages are in queue (increments when enqueue, decrements when received from queue). It turned out that the queue was empty when "try_send" was called but still the method failed.

I looked a bit at the API and it seems like the only thing that can make "try_send" fail is when the queue points to null? That shouldn't happen when the queue is empty, right?

I have 2 questions for the queue.
1. Have you tried to test using it in an interrupt that occurs frequently?
2. I did notice that there are some std library calls in the queue part of the code but I would expect those to be very heavy. Would you think it's in general a better idea to not use queue within interrupt?

Thanks.

Zack

**Comment by ilg-ul** [comment]
> safe to use the queue API within interrupt?

the `try_send()` should be ok.

> when the interrupt fires at 3~5Hz, the "try_send" method would just stop working at some point.

did you check the status code returned by `try_send()`? 

> I looked a bit at the API and it seems like the only thing that can make "try_send" fail is when the queue points to null? That shouldn't happen when the queue is empty, right?

there are several conditions, `EINVAL`, `EMSGSIZE` and `EWOULDBLOCK`.

> Have you tried to test using it in an interrupt that occurs frequently?

it passed the CMSIS RTOS validation tests, but I'm not sure that it covers this.

> I did notice that there are some std library calls in the queue part of the code but I would expect those to be very heavy. 

I took a look and I noticed only `memcpy()`, which is as heavy as the messages are.

which one do you think can be problematic?

> Would you think it's in general a better idea to not use queue within interrupt?

well, it depends. by definition queues pass messages, which means you need to copy memory into and out of the queue. this inherently takes some time.

in some cases you can arrange for the buffers to be separate, and pass only pointers to them via the queue, which is as fast as it can be.

**Comment by ilg-ul** [comment]
FYI, a friend confirmed successfully using queues with 130 bytes/msg at 200 Hz, without any problems, so I would say that passing small messages from ISRs should generally be ok.

**Comment by quzhi010** [comment]
Thanks for the response. I think I was mixing something else up with the queue API. It looks like the real issue is that some interrupts were blocked by the core and didn't get serviced.

So in the code, if an interrupt wants to use os API, it would need to be of lower priority than the core. However, I want to get interrupt for each UART byte I receive and it looks like I'm missing some of the interrupt. Since I keep the interrupt enabled throughout my software, my feeling is that the os core sometimes blocks it and that's why I'm missing some of the interrupt.

I'm not exactly sure whether this is common for all os? If I want to have an interrupt with higher priority than os core, is there a way I can still use some info from that interrupt to wake up a thread? The reason I originally set it lower was because the try_send API requires it. Is there a way around it here?

Thanks.

Zack

**Comment by quzhi010** [comment]
Also, is there any place in the os code that completely disables interrupt? I tried to bump some interrupts to highest level but it seems like it can still get blocked.

I noticed the interrupts are set to be nested and my highest interrupt priority interrupt still doesn't get to run, even when I just removed all other user interrupts. However, I wasn't able to find a place in the code that specifically does a global interrupt disable either. Can you share some comment on this? Is interrupt disabled for some portion of the time? If so, what is the expected way to handle external interrupts that occur during that period?

Thanks.

Zack

**Comment by ilg-ul** [comment]
> my feeling is that the os core sometimes blocks it and that's why I'm missing some of the interrupt.

the os shortly disables interrupts on most system calls, and while processing systick interrupts, but this is should not impact usarts with reasonable baud rates.

> I'm not exactly sure whether this is common for all os?

it is, all schedulers need to keep track of their resources (like the list of ready to run threads, timeouts, mutexes, semaphores, etc). inserting/removing from such lists requires short critical sections which prevent concurent uses.

> If I want to have an interrupt with higher priority than os core, is there a way I can still use some info from that interrupt to wake up a thread?

from fast interrupts it is not posible to call system functions, since they can occur at any time, including during the system critical sections.

you need to figure out at application level. fast interrupts are usually used to store the incomming data, or transmit the outgoing data, but from them you need to trigger a regular (non-fast) interrupt and from there you can notify the system.

> is there any place in the os code that completely disables interrupt?

if the mcu provides interrupt thresholds, the interrupt critical sections always raise the interrupt level to the desired level, leaving fast interrupts enabled.

> Is interrupt disabled for some portion of the time?

I don't have recent figures at hand, but on previous versions of µOS++ running on avr-8 at 10 MHz with the systick at 1000 Hz, the total system overhead was less than 5 %, which means 50 µS at every millisecond. for modern cortex-m3/4 running much faster, I expect this overhead to be even lower, probably 2-3%. the interrupt critical sections that occur are obviously shorter than the total 20-30 µS, a few µS I would say.

an usart at 115200 baud generates less than 11520 interrupts/second, 11520/1000000µS, one each 86 µS, so the system critical sections should not prevent them to be processed, not to mention that modern peripherals have an internal fifo to buffer incomming bytes.

if you want to process much faster usarts, probably you should consider dma, and trigger an interrupt only when the full buffer is completed.




if the mcu does not provide thresholds, the only way to implement critical sections is to completely disable interrupts.

> I wasn't able to find a place in the code that specifically does a global interrupt disable either.

I don't remember such a place, but perhaps a full search can help. check both the rtos core and the cortex-m specific code. the function to disable interrupts should be in the cortex-m package, get the function name and search for it.

**Comment by quzhi010** [comment]
Thanks for the detailed explanation.

The reason I was asking about the interrupt was because I was trying to get the boundary of the system. I'm sure there are ways around any issue I can see but it would be ideal if I can understand the issue so the system would be more stable.

But back to the missing interrupt problem, I'm actually using a micro with M4 core (STM32F405) and the code doesn't explicitly disable interrupt at any point. That's why I was struggling to understand why I was missing some interrupts and was trying to see whether the os was doing it.

I did do the global search and didn't find any place that disables interrupt for all. However, I was able to pin down the issue after some investigation. It turns out that the UART character I sent from my PC wasn't always being properly received by the UART module on the board. When I tried to do a loop back on the same board, everything works out as expected.

Sorry for blaming this on the OS. I was just trying to see where the boundary of the system is. :)

Thanks again for the help.

**Comment by ilg-ul** [comment]
You're welcome! 

As you could see, the boundaries are quite far, anyway farther than expected.
